client/packages/core/src/Reactor.js

```diff
@@ -583,25 +593,81 @@ export default class Reactor {
    *  Returns an unsubscribe function
    */
   subscribeQuery(q, cb) {
+    return this._subscribeQuery(q, cb, false);
+  }
+
+  _subscribeQuery(q, cb, once) {
     const eventId = uuid();
     const hash = weakHash(q);
 
+    if (!once) {
+      const prevResult = this.getPreviousResult(q);
+      if (prevResult) {
+        cb(prevResult);
+      }
+    }
+
     this.queryCbs[hash] = this.queryCbs[hash] || [];
-    this.queryCbs[hash].push(cb);
+    this.queryCbs[hash].push({ q, cb, once: Boolean(once) });
     this.querySubs.set((prev) => {
       prev[hash] = prev[hash] || { q, result: null, eventId };
       return prev;
     });
     this._trySendAuthed(eventId, { op: "add-query", q });
-    const prevResult = this.getPreviousResult(q);
-    if (prevResult) {
-      cb(prevResult);
-    }
+
     return () => {
-      this.queryCbs[hash] = this.queryCbs[hash].filter((x) => x !== cb);
+      this._unsubQuery(q, hash, cb);
     };
   }
 
+  async queryOnce(q) {
+    if (!this._isOnline) {
+      throw new Error(
+        "Offline: Cannot execute query because the device is offline.",
+      );
+    }
+
+    const result = await new Promise((resolve, reject) => {
+      let done = false;
+
+      const unsub = this._subscribeQuery(
```

Comment:
is there a way we can remove the `once` check? 

Maybe we can follow a different codepath from `_subscrubeQuery` or something. 

I would really love to avoid it if possible